src/internal/coredb/projects.go

```diff
@@ -0,0 +1,176 @@
+package coredb
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+// ProjectIterator batches a page of projectRow entries. Entries can be retrieved using iter.Next().
+type ProjectIterator struct {
+	lock     sync.Mutex
+	pageNum  int
+	projects []projectRow
+	index    int
+	db       *pachsql.DB
+}
+
+type projectRow struct {
+	Name        string    `db:"name"`
+	Description string    `db:"description"`
+	CreatedAt   time.Time `db:"created_at"`
+}
+
+// Next advances the iterator by one row. It returns an io.EOF when there are no more entries.
+func (iter *ProjectIterator) Next(ctx context.Context, project *pfs.ProjectInfo) error {
+	if project == nil {
+		return errors.Wrap(fmt.Errorf("project is nil"), "failed to get next project")
+	}
+	var err error
+	iter.lock.Lock()
+	defer iter.lock.Unlock()
+	if iter.index >= len(iter.projects) {
+		iter.index = 0
+		iter.pageNum++
+		iter.projects, err = listProjectPage(ctx, iter.db, 100, iter.pageNum)
+		if err != nil {
+			return errors.Wrap(err, "failed to list project page")
+		}
+		if len(iter.projects) == 0 {
+			return io.EOF
+		}
+	}
+	row := iter.projects[iter.index]
+	projectTimestamp, err := types.TimestampProto(row.CreatedAt)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert time.Time to proto timestamp")
+	}
+	*project = pfs.ProjectInfo{
+		Project:     &pfs.Project{Name: row.Name},
+		Description: row.Description,
+		CreatedAt:   projectTimestamp}
+	iter.index++
+	return nil
+}
+
+// ListProject returns a ProjectIterator that exposes a Next() function for retrieving *pfs.ProjectInfo references.
+func ListProject(ctx context.Context, db *pachsql.DB) (*ProjectIterator, error) {
+	page, err := listProjectPage(ctx, db, 100, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list projects")
+	}
+	iter := &ProjectIterator{
+		projects: page,
+		db:       db,
+	}
+	return iter, nil
+}
+
+func listProjectPage(ctx context.Context, db *pachsql.DB, pageSize, pageNum int) ([]projectRow, error) {
+	var page []projectRow
+	if err := db.SelectContext(ctx, &page, "SELECT name,description,created_at FROM core.projects ORDER BY id ASC LIMIT $1 OFFSET $2",
+		pageSize, pageSize*pageNum); err != nil {
```

Comment:
I'm not sure I like doing this multiplication here.  I'd call the args "offset" and "limit" and let the caller figure this out.  It's the most clear 1:1 mapping with the underlying data storage.